### PR TITLE
Street Fighter 2 Turbo implementation

### PR DESCRIPTION
**Pull request by Jaewongtongsoup**

### CODE REVIEW

File: Games/Retro_ML.StreetFighter2Turbo/Neural/Scoring/CombatScoreFactor.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            currScore += (currTimer * TimerInfluence + (1.0 - TimerInfluence)) * ScoreMultiplier * hpDelta;
+        }
+
+        public IScoreFactor Clone()
+        {
+            return new CombatScoreFactor()
+            {
+                ScoreMultiplier = ScoreMultiplier,
+                TimerInfluence = TimerInfluence,
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/Neural/Scoring/CombatScoreFactor.cs:81]: Missing IsDisabled = IsDisabled
```suggestion
                ScoreMultiplier = ScoreMultiplier,
                IsDisabled = IsDisabled,
                TimerInfluence = TimerInfluence,
```

File: Games/Retro_ML.StreetFighter2Turbo/Neural/Scoring/EndRoundScoreFactor.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                    currScore += ScoreMultiplier * DrawMultiplier;
+                }
+            }
+        }
+
+        public void LevelDone()
+        {
+            shouldStop = false;
+            maxTime = 0;
+            isInited = false;
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/Neural/Scoring/EndRoundScoreFactor.cs:103]: Now useless
```suggestion
```

File: Games/Retro_ML.StreetFighter2Turbo/Neural/Train/SF2TExperimentFactory.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+using Retro_ML.Emulator;
+using SharpNeat.Experiments;
+using SharpNeat.NeuralNets;
+using System.Text.Json;
+
+namespace Retro_ML.StreetFighter2Turbo.Neural.Train
+{
+    internal class SF2TExperimentFactory : INeatExperimentFactory
+    {
+        public string Id => "smb-experiment-factory";
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/Neural/Train/SF2TExperimentFactory.cs:11]:
```suggestion
        public string Id => $"{appConfig.GamePluginName}-experiment-factory";
```

File: Games/Retro_ML.StreetFighter2Turbo/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
+# Super Mario Kart Plugin
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/README.md:1]: Whole file is just copied over from Super Mario Bros. Update for Street Fighter

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackType.Address, 0x02);
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Kicking());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x0A);
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackType.Address, 0x01);
+            dataFetcher!.NextFrame();
+            Assert.False(dataFetcher!.IsPlayer2Kicking());
+        }
+
+        public void IsPlayer1Throwing()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:415]:
```suggestion
        [Test]
        public void IsPlayer1Throwing()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackType.Address, 0x00);
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Punching());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x0A);
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackType.Address, 0x01);
+            dataFetcher!.NextFrame();
+            Assert.False(dataFetcher!.IsPlayer2Punching());
+        }
+
+        public void IsPlayer1Kicking()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:388]:
```suggestion

        [Test]
        public void IsPlayer1Kicking()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackStrength.Address, new byte[] { 0x00, 0x02 });
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Throwing());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x0A);
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.AttackStrength.Address, new byte[] { 0x02, 0x04 });
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Throwing());
+        }
+
+        public void GetPlayer1AttackStrength()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:442]:
```suggestion

        [Test]
        public void GetPlayer1AttackStrength()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+            mockEmulatorAdapter!.SetMemory(Addresses.Player1Addresses.AttackType.Address, new byte[] { 0x04, 0x04 });
+            dataFetcher!.NextFrame();
+            Assert.AreEqual(1.0, dataFetcher!.GetPlayer1AttackStrength());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player1Addresses.State.Address, 0x0A);
+            mockEmulatorAdapter!.SetMemory(Addresses.Player1Addresses.AttackType.Address, new byte[] { 0x05, 0x05 });
+            dataFetcher!.NextFrame();
+            Assert.AreEqual(1.0, dataFetcher!.GetPlayer1AttackStrength());
+        }
+
+        public void GetPlayer2AttackStrength()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:479]:
```suggestion

        [Test]
        public void GetPlayer2AttackStrength()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+        [Test]
+        public void IsPlayer2Attacking()
+        {
+            Assert.False(dataFetcher!.IsPlayer2Attacking());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x0A);
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Attacking());
+        }
+
+        public void IsPlayer1Punching()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:361]:
```suggestion

        [Test]
        public void IsPlayer1Punching()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+            Assert.False(dataFetcher!.IsPlayer2Staggered());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x14);
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Staggered());
+            mockEmulatorAdapter!.SetMemory(Addresses.Player2Addresses.State.Address, 0x0E);
+            dataFetcher!.NextFrame();
+            Assert.True(dataFetcher!.IsPlayer2Staggered());
+        }
+
+        public void IsPlayer1Attacking()
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:344]:
```suggestion

        [Test]
        public void IsPlayer1Attacking()
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+using Retro_ML.StreetFighter2Turbo.Game;
+using Retro_ML_TEST.Emulator;
+
+namespace Retro_ML_TEST.Game.StreetFighter2Turbo
+{
+    [TestFixture]
+    internal class DataFetcherTest
+    {
+        delegate bool TestFunction();
+
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:13]:
```suggestion
```

File: Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs [hunk 12]
```diff
@@ -0,0 +1,516 @@
+﻿using NUnit.Framework;
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/DataFetcherTest.cs:1]: After fixing the problems in this file, don't forget to run the tests to make sure none of them fail.

File: Retro_ML_TEST/Game/StreetFighter2Turbo/ScoreFactorsTest.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            emu!.SetMemory(Addresses.GameAddresses.RoundTimer.Address, 0xA2);
+            sf.Update(df!);
+            sf.LevelDone();
+            double secondScore = firstScore + (df!.GetRoundTimerNormalized() * 0.50 + (1.0 - 0.5)) * 100 * (df!.GetPlayer1HpNormalized() - df!.GetPlayer2HpNormalized());
+            Assert.AreEqual(secondScore, sf.GetFinalScore());
+        }
+    }
+
+
+}
```
Comment (Tourmi) [inline Retro_ML_TEST/Game/StreetFighter2Turbo/ScoreFactorsTest.cs:107]:
```suggestion
    }
}
```

File: Games/Retro_ML.StreetFighter2Turbo/Game/Addresses.cs [hunk 14]
```diff
@@ -0,0 +1,141 @@
+﻿namespace Retro_ML.StreetFighter2Turbo.Game
+{
+    /// <summary>
+    /// RAM addresses used in Street Fighter 2 Turbo
+    ///
```
[the remark this answers is not in the document]
Comment (Jaewongtongsoup) [reply Games/Retro_ML.StreetFighter2Turbo/Game/Addresses.cs:5]: Nope !

File: Games/Retro_ML.StreetFighter2Turbo/README.md [hunk 15]
```diff
@@ -1,70 +1,74 @@
...
-* **Died**
-  * The amount of points to attribute to an AI that died. 
-  * It is recommended to set this value to a negative value to discourage AIs from killing themselves.
-* **Distance travelled**
-  * Points to attribute for each tile the AI traverses. This is based on the maximum distance, so going back and forth will not give more points.
-  * The AI must be grounded for points to be attributed, so jumping down a pit will not give extra points.
-  * East, West, Up, Down multipliers
+* **Combat**
+  * The amount of points to attribute to an AI for its combat prowess at the end of a round. 
+  * The amount of points depends on the time of the fight and the HP delta between the AI and the enemy.
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/README.md:21]: Don't forget to describe what the TimerInfluence does

File: Games/Retro_ML.StreetFighter2Turbo/README.md [hunk 16]
```diff
@@ -1,70 +1,74 @@
...
-* **Won level**
-  * The amount of points to attribute if the AI wins a level. Ideally, this should be a high value to encourage actually finishing levels.
-* **Coins**
-  * The amount of points to give the AI per coin collected.
-* **Power Up**
-  * Points given whenever the AI collects a power up it didn't have yet.
-  * Mushroom, Flower multipliers, Stars and 1-up
-    * Additional multiplier for the type of power-up collected. If the global multiplier is 4, and the star multiplier is 2, then the AI is awarded 8 points for collecting a star.
+* **Stopped Fighting**
+  * Stops the current fight if the AI has not hit the enemy for a specific amount of frames. Used to speed up the training since we want the ai to be on the offensive the most he can.
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/README.md:26]:
```suggestion
  * Stops the current fight if the AI has not hit the enemy for a specific amount of frames. Used to speed up training by skipping AIs that don't fight.
```

File: Games/Retro_ML.StreetFighter2Turbo/README.md [hunk 17]
```diff
@@ -1,70 +1,74 @@
...
-  * `Up` : Allows climbing on vines.
-  * `Down` : Allows crouching if mario is in Big state. Also allows Mario to go down pipes or in water levels.
+  * `A` : Middle Kick
+  * `B` : Low Kick
+  * `X` : Middle Punch
+  * `Y` : Low Punch
+  * `Left Shoulder` : High Punch
+  * `Right Shoulder` : High Kick
+  * `Left` : Move Left / Block
+  * `Right` : Move Right
```
Comment (Tourmi) [inline Games/Retro_ML.StreetFighter2Turbo/README.md:71]:
```suggestion
  * `Left` : Move Left, blocks if moving away from enemy
  * `Right` : Move Right, blocks if moving away from enemy
```

**Review by Tourmi — CHANGES_REQUESTED** [review]

**Review by Tourmi — APPROVED** [review]

**Pull request merged**